connectors/es/document.py

```diff
@@ -0,0 +1,45 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+
+from connectors.logger import logger
+
+
+class InvalidDocumentSourceError(Exception):
+    pass
+
+
+class ESDocument:
```

Comment:
This block can be greatly simplified like this:

```python
class ESDocument:
    def __init__(self, elastic_index, doc_source):
        self.index = elastic_index
        self.id = doc_source.get("_id")
        if not isinstance(self.id, str):
            raise InvalidDocumentSourceError(f"Invalid format found for id: {self.id}")

        self._source = doc_source.get("_source", {})
        if not isinstance(self._source, dict):
            raise InvalidDocumentSourceError(
                f"Invalid format found for source: {self._source}"
            )
```

Reply:
The reason I wrap it with `try... except`, is to handle edge case where some funny value is passed in for `doc_source`, e.g. `None`, or a string like `"foobar"`

Reply:
This is exactly what I proposed initially: https://github.com/elastic/connectors-python/pull/460#discussion_r1105347125, and why I didn't go that way: https://github.com/elastic/connectors-python/pull/460#discussion_r1105420783

I'm also not trying to check `isinstance(doc_source, ObjectApiResponse)`, as I think if a `dict` is provided as the `doc_source`, it should work as well.

Reply:
you should still try to be explicit because having different types allowed for  `doc_source` makes the instanciation of this class a bit fuzzy.

If you want to support several types as inputs, what is usually done is the following:

```
class ESDocument:
    def __init__(self, elastic_index, doc_source):
        self.index = elastic_index

        if not isinstance(doc_source, dict):            
            raise InvalidDocumentSourceError(f"Invalid format found for doc_source: {str(doc_source)}")
 
        self.id = doc_source.get("_id")
        if not isinstance(self.id, str):
            raise InvalidDocumentSourceError(f"Invalid format found for id: {self.id}")

        self._source = doc_source.get("_source", {})
        if not isinstance(self._source, dict):
            raise InvalidDocumentSourceError(
                f"Invalid format found for source: {self._source}"
            )

    @classmethod 
    def from_api_response(cls, api_response):
        return cls(api_response.body)    
```

Reply:
`ObjectApiResponse` is defined in `elastic-transport`, and has method `__getitem__` defined:
https://github.com/elastic/elastic-transport-python/blob/6d6e64ae26a4088801f6df1f3bdb40fea88b9b75/elastic_transport/_response.py#L177-L188

The reason why I don't want to use `isinstance(doc_source, ObjectApiResponse)`:
1. It should work for `dict` as well (actually any object with method `get`)
2. I don't want to add dependency on `elastic_transport` here

Or we could simply do this: https://github.com/elastic/connectors-python/pull/460#discussion_r1106263366, which will raise any possible exception, instead of `InvalidDocumentSourceError`

Reply:
> It should work for dict as well (actually any object with method get)

Can you elaborate on this?  Do you mean it should support the mapping interface? Will we really use this class for any kind of object ? 

> I don't want to add dependency on elastic_transport here

In that case you can check for `get` presence:

```python
if not hasattr(data_source, 'get'):
    raise InvalidDocumentSourceError("Expected a class with get")
```

But it would be best to be explicit about what the class expects IMHO. 
You could also make it compatible only with dict and ask the caller to take care of passing a dict, so you separate concerns. THe code that manipulates `ObjectApiResponse` could simply do:

```
ESDocument(index, my_response.body)
```

Reply:
>Can you elaborate on this?

If a caller wants to instantiate an `ESDocument` object, he/she can pass in an `ObjectApiResponse` object (as what we do in `connectors-python`:
```python
resp = a_client.get(index=".elastic-connectors`, id="1") # an ObjectApiResponse object
doc = ESDocument(an_index, resp)
```
he/she can also pass in a dict:
```python
connector_doc = {
    "_id": "1",
    "_source": {}
}
doc = ESDocument(an_index, connector_doc)
```

Or any object with `get` method.

But again, `YAGNI`.

The reason why I don't want to handle `ObjectApiResponse` explicitly, it's that I'm hesitate to have `ObjectApiResponse` imported as a dependency:
```python
# _response module starts with _ implies it's not supposed to be used directly?
from elastic_transport._response import ObjectApiResponse

class ESDocument:
    def __init__(self, elastic_index, doc_source):
        self.index = elastic_index
        if not isinstance(doc_source, ObjectApiResponse):
            raise InvalidDocumentSourceError(
                f"Invalid doc_source type found: {doc_source.__class__.__name__}, expected: {ObjectApiResponse.__class__.__name__}"
            )
        ...
```

Reply:
But what is the intent of `ESDocument` -- can you add a two-sentence description in its docstring?

Why a `dict` or an `ObjectApiResponse`?
Will it ever be instantiated with something else than a document from an Elasticsearch client? 
It seems to me that we will never use that class with different type of data.

In any case, if you want both, you don't need to import `elastic_transport`, what prevents you to do: 

```
resp = a_client.get(index=".elastic-connectors`, id="1") # an ObjectApiResponse object
doc = ESDocument(an_index, resp.body)
```
Because your class just needs the dict from the body and that's what `ObjectApiResponse.body` returns

Reply:
>Will it ever be instantiated with something else than a document from an Elasticsearch client?

No, not for now.

> In any case, if you want both, you don't need to import `elastic_transport`, what prevents you to do

Will update.

Reply:
Done here: https://github.com/elastic/connectors-python/pull/460/commits/f7bb6fd9bb919caaad753682cb8335172a60b55e